% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/single_molecule_manipulation_functions.r
\name{BinMethylation}
\alias{BinMethylation}
\title{Summarize methylation inside sorting bins}
\usage{
BinMethylation(MethSM, TFBS, bin)
}
\arguments{
\item{MethSM}{Single molecule matrix}

\item{TFBS}{Transcription factor binding site to use for sorting, passed as a GRanges object of length 1}

\item{bin}{vector of two integers representing the coordinate of a bin relative to the center of the TFBS}
}
\value{
Reads covering bin with their summarized methylation status
}
\description{
Summarize methylation inside sorting bins
}
\examples{
Qinput = paste0(tempdir(), "/NRF1Pair_Qinput.txt")
library(BSgenome.Mmusculus.UCSC.mm10)

if(file.exists(Qinput)){
    QuasRprj = GetQuasRprj(Qinput, BSgenome.Mmusculus.UCSC.mm10)

    MySample = readr::read_delim(Qinput, delim = "\t")$SampleName[1]
    Region_of_interest = GRanges(seqnames = "chr6", ranges = IRanges(start = 88106000, end = 88106500), strand = "*")

    Methylation = CallContextMethylation(sampleSheet = Qinput,
                                     sample = MySample,
                                     genome = BSgenome.Mmusculus.UCSC.mm10,
                                     range = Region_of_interest,
                                     coverage = 20,
                                     ConvRate.thr = 0.2)

    TFBSs = GenomicRanges::GRanges("chr6", IRanges(c(88106253), c(88106263)), strand = "-")
    elementMetadata(TFBSs)$name = c("NRF1")
    names(TFBSs) = c(paste0("TFBS_", c(4305216)))

    binMethylationValues = BinMethylation(MethSM = Methylation[[2]], TFBS = TFBSs, bin = c(-15,15))
}

}
